### PR TITLE
Support for recorded and paused time - Modifications after 2nd code review

### PR DESCRIPTION
**Pull request by FJBDev**

### CODE REVIEW

File: bundles/net.tourbook.statistics/src/net/tourbook/statistics/graphs/StatisticDay.java [hunk 1]
```diff
@@ -301,6 +301,7 @@ private ChartToolTipInfo createToolTipData(int valueIndex) {
       final ZonedDateTime zdtTourStart = _tourDayData.allStartDateTimes.get(valueIndex);
       final ZonedDateTime zdtTourEnd = zdtTourStart.plusSeconds(elapsedTime);
 
+      //TODO FB propose pace based on RECORDED TIME ?
```
Comment (FJBDev) [inline bundles/net.tourbook.statistics/src/net/tourbook/statistics/graphs/StatisticDay.java:304]: @wolfgang-ch 
What do you think ? should pace be computed only with elapsed time ? Should we give the ability for the user to select which time is used to compute the pace in the preferences ? (Elapsed vs Recorded vs Moving)

File: bundles/net.tourbook/src/net/tourbook/map2/view/TourMapPainter.java [hunk 2]
```diff
@@ -1351,7 +1351,7 @@ private boolean doPaint_Pauses(final GC gcTile,
 
       if (tourData.isMultipleTours()) {
 
-         //TODO FB support joining pauses when creating a multiple tour
+         //TODO FB support joining pauses when creating a multiple tour, what is a multiple tour ?
```
Comment (FJBDev) [inline bundles/net.tourbook/src/net/tourbook/map2/view/TourMapPainter.java:1354]: @wolfgang-ch I added the support to merge the pauses when concatenating tours but I don't know what a "multiple" tour is ? should the pauses be merged for multiple tours ?

File: bundles/net.tourbook.statistics/src/net/tourbook/statistics/graphs/StatisticDay.java (hunk 1 above)
Comment (wolfgang-ch) [reply bundles/net.tourbook.statistics/src/net/tourbook/statistics/graphs/StatisticDay.java:304]: I'm mainly interrested in speed/pace values which are based on the moving time, the columns in tour book view are based on this value but as a new feature you can provide the user the option to select the base time.

mt-speed

File: bundles/net.tourbook/src/net/tourbook/map2/view/TourMapPainter.java (hunk 2 above)
Comment (wolfgang-ch) [reply bundles/net.tourbook/src/net/tourbook/map2/view/TourMapPainter.java:1354]: You have not yet used this feature ?

Multiple tours can be selected e.g. in the tourbook view.

mt-multiple-tours

**Comment by wolfgang-ch** [comment]
I like the new time icons :-)

mt-time-icons

I've not yet merged into `main`, first I'll do more tests

**Comment by wolfgang-ch** [comment]
I've found/fixed an SQL exception https://github.com/wolfgang-ch/mytourbook/commit/2281360a4ce06c6b4cff499de16f9334962319b9
and changed a mnemonic https://github.com/wolfgang-ch/mytourbook/commit/c20499127e8024a78dce21376e91fb74673238af

These modifications are in your FJBDev branch

**Comment by FJBDev** [comment]
> 
> 
> I like the new time icons :-)
> 
> mt-time-icons
> 
> I've not yet merged into `main`, first I'll do more tests

Thanks, I like them too!
But i am still not sure that I can use them as they [don't work in Linux](https://github.com/wolfgang-ch/mytourbook/pull/222#discussion_r477431719) :-(

**Comment by FJBDev** [comment]
> 
> 
> I've found/fixed an SQL exception [2281360](https://github.com/wolfgang-ch/mytourbook/commit/2281360a4ce06c6b4cff499de16f9334962319b9)
> and changed a mnemonic [c204991](https://github.com/wolfgang-ch/mytourbook/commit/c20499127e8024a78dce21376e91fb74673238af)
> 
> These modifications are in your FJBDev branch

Thank you, i just looked at them, that looks good to me.

File: bundles/net.tourbook/src/net/tourbook/map2/view/TourMapPainter.java (hunk 2 above)
Comment (FJBDev) [reply bundles/net.tourbook/src/net/tourbook/map2/view/TourMapPainter.java:1354]: Oh you're saying that a multiple tour is when you select several tours in the tourbook view ? Then yes, I have used this feature but didn't know it had a name :-)
I thought a multiple tour was a tour that was merged.

File: bundles/net.tourbook.statistics/src/net/tourbook/statistics/graphs/StatisticDay.java (hunk 1 above)
Comment (FJBDev) [reply bundles/net.tourbook.statistics/src/net/tourbook/statistics/graphs/StatisticDay.java:304]: Ok, I will give some thought to that.

**Comment by FJBDev** [comment]
@wolfgang-ch  That's what happens in Linux 

image

**Comment by wolfgang-ch** [comment]
Ubuntu is OK

mt-icons-ubuntu

**Comment by wolfgang-ch** [comment]
New time icon sugesstions:

* Replaced elapsed icon with recorded icon
* New recorded icon
* Some suggestions for the moving icon. When searched for a new moving icon, I found this document https://www.unicode.org/L2/L2018/18090-tachograph-symbols.pdf, it is interresting to see for their process how to find a new icon, I googled for "unicode tacho"

mt-time-symbols-ubuntu-2


mt-time-symbols

Which is your preferred moving icon?

**Comment by wolfgang-ch** [comment]
Any unicode character could be used

mt-moving-icons-II

mt-moving-icons-II-win10


Selecting a character is easy with LibreOffice

mt-icons

**Comment by wolfgang-ch** [comment]
I've a lot of tours with overlapped paused times at the beginning of the tour, zooming in do not help to read the displayed values

mt-paused-time-in-tour-chart

The tour segmenter shows stacked labels which prevents overlapping, this is done in `net.tourbook.ui.tourChart.ValueOverlapChecker` but the whole is a relative complex code :-)

mt-tour-segmenter-with-stacked-labels

**Comment by wolfgang-ch** [comment]
Some tours have negative paused time

mt-negative-paused-time

**Comment by FJBDev** [comment]
> 
> 
> Ubuntu is OK
> 
> mt-icons-ubuntu

Maybe it depends the desktop used. Are you using [Gnome](https://en.wikipedia.org/wiki/GNOME) ? I am using [Cinnamon](https://en.wikipedia.org/wiki/Cinnamon_(desktop_environment))

**Comment by FJBDev** [comment]
In GNOME, 
` In GNOME 3, the default font is Cantarell,` Source : [https://developer.gnome.org/hig/stable/typography.html.en](https://developer.gnome.org/hig/stable/typography.html.en)

**Comment by FJBDev** [comment]
And in my Linux Debian 10.5 Cinnamon, this is what I have as a default font : 
image

**Comment by FJBDev** [comment]
> 
> 
> New time icon sugesstions:
> 
>     * Replaced elapsed icon with recorded icon
> 
>     * New recorded icon
> 
>     * Some suggestions for the moving icon. When searched for a new moving icon, I found this document https://www.unicode.org/L2/L2018/18090-tachograph-symbols.pdf, it is interresting to see for their process how to find a new icon, I googled for "unicode tacho"
> 
> 
> mt-time-symbols-ubuntu-2
> 
> mt-time-symbols
> 
> Which is your preferred moving icon?

My preferred moving icon is the one in the red square: 

image

**Comment by FJBDev** [comment]
> 
> 
> Any unicode character could be used

Not if it doesn't work on all the Linux :anguished:

**Comment by FJBDev** [comment]
> net.tourbook.ui.tourChart.ValueOverlapChecker

I have sen this issue as well. I will look into this complex code `net.tourbook.ui.tourChart.ValueOverlapChecker`

**Comment by FJBDev** [comment]
> 
> 
> Some tours have negative paused time
> 
> mt-negative-paused-time

Are those tours old tours reimported ? are they new tours ? Which format are they ? FIT ? TCX ?
I have never seen this issue, this is not normal.

it's probably because the elapsed time is 1 second less than the recorded time:
image

My guess is that it's from a FIT file but the elapsed value should NOT be less than the recorded value. 
I will fix that in my code but you will need to verify because I dont have such FIT files.

**Comment by FJBDev** [comment]
See my fix attempt [here](https://github.com/wolfgang-ch/mytourbook/pull/228).
Does it fix your paused time of -1 issue ?

**Comment by wolfgang-ch** [comment]
> And in my Linux Debian 10.5 Cinnamon, this is what I have as a default font :

I'm using Ubuntu 20.04 with it's default setup

ubuntu

**Comment by wolfgang-ch** [comment]
> My preferred moving icon is the one in the red square:
> 
> image

Fixed

**Comment by wolfgang-ch** [comment]
> > Any unicode character could be used
> 
> Not if it doesn't work on all the Linux 😧

I would understand that 3-Byte Unicode characters, e.g. all of these moving time icons

mt-moving-icons-II

are not supported but you have choosen only 2-Byte Unicode characters

**Comment by wolfgang-ch** [comment]
> 
> 
> > Some tours have negative paused time
> 
> Are those tours old tours reimported ? are they new tours ? Which format are they ? FIT ? TCX ?
> I have never seen this issue, this is not normal.
> 
> it's probably because the elapsed time is 1 second less than the recorded time:
> image
> 
> My guess is that it's from a FIT file but the elapsed value should NOT be less than the recorded value.
> I will fix that in my code but you will need to verify because I dont have such FIT files.

This issue occurred with FIT files, imported and reimported

**Comment by wolfgang-ch** [comment]
Statistic paused time value is larger than elapsed time value

mt-paused-time-statistic

**Comment by wolfgang-ch** [comment]
Missing mnemonics

mt-stat-options

**Comment by wolfgang-ch** [comment]
I've several tours with wrong paused time, these are mainly tours which I recorded when driving by car but I also found 1 normal bicycle tour with wrong paused time value. These tours are all recorded with an Edge 1030 -> FIT format.


mt-stat-values

**Comment by FJBDev** [comment]
> 
> 
> > > Some tours have negative paused time
> > 
> > 
> > Are those tours old tours reimported ? are they new tours ? Which format are they ? FIT ? TCX ?
> > I have never seen this issue, this is not normal.
> > it's probably because the elapsed time is 1 second less than the recorded time:
> > image
> > My guess is that it's from a FIT file but the elapsed value should NOT be less than the recorded value.
> > I will fix that in my code but you will need to verify because I dont have such FIT files.
> 
> This issue occurred with FIT files, imported and reimported

Did you try with my new code available in my latest [PR](https://github.com/wolfgang-ch/mytourbook/pull/228)?

**Comment by FJBDev** [comment]
> 
> 
> I've several tours with wrong paused time, these are mainly tours which I recorded when driving by car but I also found 1 normal bicycle tour with wrong paused time value. These tours are all recorded with an Edge 1030 -> FIT format.
> 
> mt-stat-values

I can't reproduce the issue so I'm not sure I can troubleshoot....unless you can share one or more FIT files ?

**Comment by FJBDev** [comment]
> 
> 
> > And in my Linux Debian 10.5 Cinnamon, this is what I have as a default font :
> 
> I'm using Ubuntu 20.04 with it's default setup
> 

Ok, so it shows that you are using GNOME 3, so maybe it's because GNOME and Cinnamon use a different font ?

**Comment by FJBDev** [comment]
> 
> 
> Missing mnemonics
> 
> mt-stat-options

Fixed

**Comment by FJBDev** [comment]
> 
> 
> > > Any unicode character could be used
> > 
> > 
> > Not if it doesn't work on all the Linux 😧
> 
> I would understand that 3-Byte Unicode characters, e.g. all of these moving time icons
> 
> mt-moving-icons-II
> 
> are not supported but you have choosen only 2-Byte Unicode characters

Maybe it's the default font used by GNOME and Cinnamon ?
I will continue to investigate.

File: bundles/net.tourbook.statistics/src/net/tourbook/statistics/graphs/StatisticDay.java (hunk 1 above)
Comment (FJBDev) [reply bundles/net.tourbook.statistics/src/net/tourbook/statistics/graphs/StatisticDay.java:304]: If I add a preference for the user to choose pace from moving time vs pace from recorded time, would that preference menu be a good place for a "Pace" group UI ? do you see a better place ?

image

**Comment by wolfgang-ch** [comment]
This PR is now merged into `main`, it was a long reviewing time.

Some more issues which I found:

* Unit column should show only units

mt-column-units


* One missing mnemonic

mt-mnemonic

**Comment by FJBDev** [comment]
> Some more issues which I found:
> 
>     * Unit column should show only units
> 
> 
> mt-column-units

This is what is in the official 20.8 so I thought it was normal to have the unit the same as the column header:
image

I can fix

**Comment by FJBDev** [comment]
>     * One missing mnemonic
> 
> 
> mt-mnemonic

Fixed

**Comment by wolfgang-ch** [comment]
> This is what is in the official 20.8 so I thought it was normal to have the unit the same as the column header:
> image
> 
> I can fix
It looked a bit strange when I first saw it and why I created this "issue". So we can keep it except "Recorded time" which shows a text

**Comment by FJBDev** [comment]
> It looked a bit strange when I first saw it and why I created this "issue". So we can keep it except "Recorded time" which shows a text

Whatever you preferred but actually the menu is very confusing to me because I don't understand the difference between the "Tour Format" column and the "Unit" column. It seems they are the same to me ?

image

**Comment by wolfgang-ch** [comment]
* Tour Format: how values are formatted

* Unit: when searching references then you can see where units are used

mt-columns

File: bundles/net.tourbook.statistics/src/net/tourbook/statistics/graphs/StatisticDay.java (hunk 1 above)
Comment (wolfgang-ch) [reply bundles/net.tourbook.statistics/src/net/tourbook/statistics/graphs/StatisticDay.java:304]: > If I add a preference for the user to choose pace from moving time vs pace from recorded time, would that preference menu be a good place for a "Pace" group UI ? do you see a better place ?

The location where this pref setting should be put on, depends how often it is modified. To put it into a prefpage, the user need more clicks to change it as when it is in a slidout

**Comment by wolfgang-ch** [comment]
> See my fix attempt [here](https://github.com/wolfgang-ch/mytourbook/pull/228).
> Does it fix your paused time of -1 issue ?

YES !

**Old**
image

**New**

mt-negative-pause

File: bundles/net.tourbook.statistics/src/net/tourbook/statistics/graphs/StatisticDay.java (hunk 1 above)
Comment (FJBDev) [reply bundles/net.tourbook.statistics/src/net/tourbook/statistics/graphs/StatisticDay.java:304]: True, I could put it in a slideout but I don't think the user will change that option often so it's probably OK to have it in the preferences.

File: bundles/net.tourbook/src/net/tourbook/map2/view/TourMapPainter.java (hunk 2 above)
Comment (FJBDev) [reply bundles/net.tourbook/src/net/tourbook/map2/view/TourMapPainter.java:1354]: Fixed

File: bundles/net.tourbook.statistics/src/net/tourbook/statistics/graphs/StatisticDay.java (hunk 1 above)
Comment (wolfgang-ch) [reply bundles/net.tourbook.statistics/src/net/tourbook/statistics/graphs/StatisticDay.java:304]: Pref page is OK for me

**Comment by FJBDev** [comment]
> 
> 
> I've a lot of tours with overlapped paused times at the beginning of the tour, zooming in do not help to read the displayed values
> 
> mt-paused-time-in-tour-chart
> 
> The tour segmenter shows stacked labels which prevents overlapping, this is done in `net.tourbook.ui.tourChart.ValueOverlapChecker` but the whole is a relative complex code :-)
> 
> mt-tour-segmenter-with-stacked-labels

Implemented the overlap for TourChart pauses in my recent [PR](https://github.com/wolfgang-ch/mytourbook/pull/236)

**Pull request merged**